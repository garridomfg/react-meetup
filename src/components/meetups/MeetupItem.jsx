import classes from "./MeetupItem.module.css";
import Card from "../ui/Card";

export const MeetupItem = ({ item }) => {
  const { image, title, address, description } = item;

  return (
    <li className={classes.item} data-test="meet-up-item">
      <Card>
        <div className={classes.image}>
          <img src={image} alt={title} />
        </div>
        <div className={classes.content}>
          <h3>{title}</h3>
          <address>{address}</address>
          <p>{description}</p>
        </div>
        <div className={classes.actions}>
          <button>Add to favorites</button>
        </div>
      </Card>
    </li>
  );
};
