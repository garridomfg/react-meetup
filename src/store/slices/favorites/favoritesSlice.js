import { createSlice } from "@reduxjs/toolkit";

export const favoritesSlice = createSlice({
  name: "favorites",
  initialState: {
    favorites: [],
  },
  reducers: {
    addFavorite: (state, action) => {
      state.favorites = [action.payload, ...state.favorites];
    },
    removeFavorite: (state, action) => {
      state.favorites = action.payload;
    },
  },
});

export const { addFavorite, removeFavorite } = favoritesSlice.actions;
