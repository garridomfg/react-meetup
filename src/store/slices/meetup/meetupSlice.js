import { createSlice } from "@reduxjs/toolkit";

export const meetupSlice = createSlice({
  name: "meetups",
  initialState: {
    meetups: [],
  },
  reducers: {
    setMeetups: (state, action) => {
      state.meetups = action.payload;
    },
    addMeetup: (state, action) => {
      state.meetups = [action.payload, ...state.meetups];
    },
  },
});

export const { setMeetups, addMeetup } = meetupSlice.actions;
