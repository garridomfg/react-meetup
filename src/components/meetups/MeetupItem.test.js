/* eslint-disable testing-library/no-debugging-utils */
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import { favoritesSlice } from "../../store/slices/favorites/favoritesSlice";
import { meetupSlice } from "../../store/slices/meetup/meetupSlice";
import { MeetupItem } from "./MeetupItem";
import { screen, render } from "@testing-library/react";

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: () => (fn) => fn(),
}));

const store = configureStore({
  reducer: {
    meetups: meetupSlice.reducer,
    favorites: favoritesSlice.reducer,
  },
});

const initialState = {
  id: "m1",
  title: "This is a first meetup",
  image:
    "https://upload.wikimedia.org/wikipedia/commons/thumb/d/d3/Stadtbild_M%C3%BCnchen.jpg/2560px-Stadtbild_M%C3%BCnchen.jpg",
  address: "Meetupstreet 5, 12345 Meetup City",
  description:
    "This is a first, amazing meetup which you definitely should not miss. It will be a lot of fun!",
};

describe("<MeetupItem/>", () => {
  beforeEach(() => jest.clearAllMocks());

  test("Should match with snapshot", () => {
    const view = render(
      <Provider store={store}>
        <MeetupItem item={initialState} />
      </Provider>
    );
    expect(view).toMatchSnapshot();
  });

  test("Should show the image with the indicated URL and ALT", () => {
    render(
      <Provider store={store}>
        <MeetupItem item={initialState} />
      </Provider>
    );
    expect(screen.getByRole("img").src).toBe(initialState.image);
    expect(screen.getByRole("img").alt).toBe(initialState.title);
  });

  test("Items should have text", () => {
    render(
      <Provider store={store}>
        <MeetupItem item={initialState} />
      </Provider>
    );
    expect(screen.getByText(initialState.title)).toBeTruthy();
    expect(screen.getByText(initialState.address)).toBeTruthy();
    expect(screen.getByText(initialState.description)).toBeTruthy();
  });
});
