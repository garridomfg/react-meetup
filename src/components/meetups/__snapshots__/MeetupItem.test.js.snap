// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MeetupItem/> Should match with snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="item"
        data-test="meet-up-item"
      >
        <div
          class="card"
        >
          <div
            class="image"
          >
            <img
              alt="This is a first meetup"
              src="https://upload.wikimedia.org/wikipedia/commons/thumb/d/d3/Stadtbild_M%C3%BCnchen.jpg/2560px-Stadtbild_M%C3%BCnchen.jpg"
            />
          </div>
          <div
            class="content"
          >
            <h3>
              This is a first meetup
            </h3>
            <address>
              Meetupstreet 5, 12345 Meetup City
            </address>
            <p>
              This is a first, amazing meetup which you definitely should not miss. It will be a lot of fun!
            </p>
          </div>
          <div
            class="actions"
          >
            <button>
              Add to favorites
            </button>
          </div>
        </div>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="item"
      data-test="meet-up-item"
    >
      <div
        class="card"
      >
        <div
          class="image"
        >
          <img
            alt="This is a first meetup"
            src="https://upload.wikimedia.org/wikipedia/commons/thumb/d/d3/Stadtbild_M%C3%BCnchen.jpg/2560px-Stadtbild_M%C3%BCnchen.jpg"
          />
        </div>
        <div
          class="content"
        >
          <h3>
            This is a first meetup
          </h3>
          <address>
            Meetupstreet 5, 12345 Meetup City
          </address>
          <p>
            This is a first, amazing meetup which you definitely should not miss. It will be a lot of fun!
          </p>
        </div>
        <div
          class="actions"
        >
          <button>
            Add to favorites
          </button>
        </div>
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
